byzcoin/service.go

```diff
@@ -554,11 +555,13 @@ func (s *Service) DownloadState(req *DownloadState) (resp *DownloadStateResponse
 		s.downloadState.stop = make(chan bool)
 		nonce := binary.LittleEndian.Uint64(random.Bits(64, true, random.New()))
 		s.downloadState.nonce = nonce
+		total := make(chan int)
 		go func(ds downloadState) {
 			idStr := fmt.Sprintf("%x", ds.id)
 			db, bucketName := s.GetAdditionalBucket([]byte(idStr))
 			err := db.View(func(tx *bbolt.Tx) error {
 				bucket := tx.Bucket(bucketName)
+				total <- bucket.Stats().KeyN
```

Comment:
i think `bucket.Stats().KeyN` will give you the total number of key-value pairs in the boltdb bucket which is always higher than the total number of states in byzcoin because the intermediate nodes in the trie are also stored as key-value pairs in boltdb. so i'm not sure whether this is what you want to show

Reply:
nevermind, it should be fine because `len(KeyValues)` is actually the boltdb key-values

Reply:
Yes, and we need to transmit the whole state, because the merkle-tree retains some history of the adding and deleting of instances. So if we just transmit the instances, the created merkle-tree will have a different root!